Add capabilities for reusable blocks

noisysocks opened the pull request:
### 🔒 What this does

Fixes #3936.

Adds capabilities for the reusable block custom post type and maps them to the default WordPress roles. The capabilities are mapped as per https://github.com/WordPress/gutenberg/issues/3936#issuecomment-353251845:

> * Contributors can see and use existing blocks. The can edit/delete blocks for which they're the `post_author`. There's no draft reusable blocks, so no equivalent for them to be able to create blocks.
> * Authors can create blocks, and edit/delete blocks for which they're the `post_author`.
> * Editors and above can create/edit/delete all blocks. They can change the `post_author` for all blocks.

Here's a table to help illustrate:

|        | Admins/Editors | Authors | Contributors | Subscribers/Visitors |
| ------ | -------- | ------- | ------------ | ------------ |
| **Create** | Yes      | Yes     | No           | No           |
| **Read**   | Yes      | Yes     | Yes          | No           |
| **Update/Delete** | Yes      | Own     | Own          | No           |

### ✅ How I tested this (and how you could too)

I've added unit tests which run through the cases described by the table above.

You could also test this manually. First, I used WP-CLI to make sure I was working with a clean slate:

```
$ wp site empty
$ wp role reset --all
```

Then I created a bunch of users with different roles:

```
$ wp user create editor editor@local.test --role=editor --user_pass=password
$ wp user create author1 author1@local.test --role=author --user_pass=password
$ wp user create author2 author2@local.test --role=author --user_pass=password
$ wp user create contributor contributor@local.test --role=contributor --user_pass=password
$ wp user create subscriber subscriber@local.test --role=subscriber --user_pass=password
```

Then I tested various reusable block API operations while authenticated as different users. Here I'm using [HTTPie](https://httpie.org), but any API client would work. Make sure your WordPress has the [basic-auth plugin](https://github.com/WP-API/Basic-Auth) installed.

```
$ http -a editor:password POST local.wordpress.test/wp-json/wp/v2/blocks title='Editor block' content=Test
```

noisysocks committed:
Add capabilities for reusable blocks

Adds capabilities for creating, reading, updating and deleting reusable
blocks. The capabilities are mapped like so:

|        | Editors^ | Authors | Contributors | Subscribers* |
| ------ | -------- | ------- | ------------ | ------------ |
| Create | Yes      | Yes     | No           | No           |
| Read   | Yes      | Yes     | Yes          | No           |
| Update | Yes      | Own     | Own          | No           |
| Delete | Yes      | Own     | Own          | No           |

^ Includes administrators.
* Includes visitors that are not logged in.

noisysocks commented on lib/register.php:
```diff
@@ -401,11 +401,48 @@ function gutenberg_register_post_types() {
+	$contributor->add_cap( 'read_blocks' );
+	$contributor->add_cap( 'delete_blocks' );
+	$contributor->add_cap( 'delete_published_blocks' );
+	$contributor->add_cap( 'edit_published_blocks' );
```
Is there a more suitable hook to do this work in? I read through the relevant WordPress docs but couldn't figure out a better place to do it.

noisysocks replied:
Fixed in a206f515f83b9121ce0b497a1a9a71317b2d7699.

noisysocks commented:
@pento @brandonpayton—I think this works (well, the test cases I wrote definitely pass...) but this capabilities stuff is pretty hairy so I'd love a second pair of 👀s on this.

pento reviewed (changes requested):
Functionality looks good, just needs the performance tweak I mentioned.

carlhancock commented:
> Contributors can see and use existing blocks. The can edit/delete blocks for which they're the post_author. There's no draft reusable blocks, so no equivalent for them to be able to create blocks.

The description and table says that Contributors can edit/delete blocks for which they're the post_author.  But at the same time it says they can only see and use existing blocks and cannot create them.  If they cannot create them, why can they edit/delete them?

Here is what the permissions table is above:

  | Admins/Editors | Authors | Contributors | Subscribers/Visitors
-- | -- | -- | -- | --
Create | Yes | Yes | No | No
Read | Yes | Yes | Yes | No
Update/Delete | Yes | Own | Own | No

Shouldn't it actually be:


  | Admins/Editors | Authors | Contributors | Subscribers/Visitors
-- | -- | -- | -- | --
Create | Yes | Yes | No | No
Read | Yes | Yes | Yes | No
Update/Delete | Yes | Own | **No** | No

Example:

- User has Author role and thus the capabilities to create blocks.
- User creates a block.
- For whatever reason User is bumped down from Author to Contributor.
- User can no longer create blocks.
- But User can edit and delete blocks he created when he originally had the Author role.

Shouldn't the Contributor have strictly Read access to use blocks and no update/delete at all if they have no create capability? Even if the block was originally created by them while they had higher role access?

noisysocks commented:
Good catch, @carlhancock. The weirdness comes from the fact that there is an implicit _publish_ action contained within the _create_ action. This is because reusable blocks are immediately published when they are created. Contributors cannot publish posts or pages, and so I made it so that they cannot create reusable blocks.

It seems totally reasonable to implement your suggestion and make it so that contributors cannot update or delete blocks.

noisysocks committed:
Only call add_cap() when necessary

Avoid calling $role->add_cap() when the role already has the given
capability. This avoids an unnecessary database write.

noisysocks committed:
Prevent contributors from editing reusable blocks

Make reusable block permissions more consistent by preventing
contributors from *editing* blocks - not just creating.

noisysocks commented:
I've updated this PR. Let me know what you think! 💃

pento reviewed (approved):
🕺🏻💯👍🏻

westonruter commented on lib/class-wp-rest-blocks-controller.php:
```diff
@@ -16,6 +16,39 @@
+	public function check_read_permission( $post ) {
+		// Ensure that the user is logged in and has the read_blocks capability.
+		$post_type = get_post_type_object( $post->post_type );
+		if ( ! current_user_can( $post_type->cap->read_post, $post->ID ) ) {
```
Minor point, but you should be able to provide the `edit_post` meta capability here instead rather than looking up the primitive directly, since `map_meta_cap()` will be doing it for you later anyway. So then this line can be simply:

```php
if ( ! current_user_can( 'read_post', $post->ID ) ) {
```

westonruter replied:
If you use `read_post` (not `read_block`) then `map_meta_cap()` should automatically map it over to `read_block` (or whatever the registered post type cap gets changed to).

EarthmanWeb commented:
Hello, @noisysocks  I'm a bit confused - It looks like this PR was merged, but I can't find these 'block' caps in WordPress core - how do I use these capabilities please?
